website/public_adventures.py

```diff
@@ -31,67 +31,168 @@ def __init__(self, db: Database, achievements: Achievements):
-                    self.adventures[_level].append(current_adventure)
-                else:
-                    self.adventures[_level] = [current_adventure]
-
-            available_levels = adventure["levels"] if adventure.get("levels") else [adventure["level"]]
-            self.customizations["available_levels"].update([int(adv_level) for adv_level in available_levels])
-
     @route("/", methods=["GET"])
     @route("/filter", methods=["POST"])
     @requires_teacher
-    def filtering(self, user, index_page=False):
+    def filtering(self, user):
+        level = request.args["level"] if request.args.get("level") else "1"
+        language = request.args.get("lang", g.lang)
+        tag = request.args.get("tag", "")
+        search = request.form.get("search", request.args.get("search", ""))
+        print('\n\n\n', level, language, tag, search, '\n\n\n')
+
+        available_languages = set()
+        available_levels = set()
+        available_tags = set()
+
+        # Get all possible filters
+        lang_level_filters = self.db.get_public_adventures_filters({"field": "lang#level", })
+        for record in lang_level_filters:
+            lang, _level = record.get("value", "#").split("#")
+            available_levels.update([_level])
+            if level == _level:
+                available_languages.update([lang])
+        customizations = {"available_levels": available_levels}
+
+        tag_filters = self.db.get_public_adventures_filters({"field": "tag"})
+        for record in tag_filters:
+            _tag = record.get("value")
+            available_tags.update([_tag])
+
+        customizations = {"available_levels": available_levels}
+
+        tags = []
+        if tag:
+            toReset = request.args.get("reset")
+            if toReset:
+                # then it's the current selected tags, so remove current's tag
+                tags = [_tag for _tag in toReset.split(",") if _tag != tag]
+            else:
+                tags = tag.split(",")
+            tags = [t for t in tags if t]
+
+        # Get indexes
+        level_adventure_ids = self.db.get_public_adventures_indexes({"field#value": f"level#{level}"})
+        lang_adventure_ids = self.db.get_public_adventures_indexes({"field#value": f"lang#{language}"})
+        tag_adventure_ids = []
+        for _t in tags:
+            tag_adventure_ids += self.db.get_public_adventures_indexes({"field#value": f"tag#{_t}"})
+
+        level_adventure_ids = [record["date#adventure_id"].split("#")[1] for record in level_adventure_ids]
+        lang_adventure_ids = [record["date#adventure_id"].split("#")[1] for record in lang_adventure_ids]
+        tag_adventure_ids = [record["date#adventure_id"].split("#")[1] for record in tag_adventure_ids]
+
+        # Filter out based on level
+        lang_adventure_ids = [_id for _id in lang_adventure_ids if _id in level_adventure_ids]
+        tag_adventure_ids = [_id for _id in tag_adventure_ids if _id in level_adventure_ids]
```

Comment:
If you make them sets of IDs, you can use set intersection. And if you use set intersection before splitting the date and the ID, you can afterwards still sort the final list (descending) to come up with a list with the newest program first.

But again, you won't be able to paginate if you do that. How big is your data?